contracts/implementation/SMAOracle.sol

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.7;
+import "../interfaces/IOracleWrapper.sol";
+import "../interfaces/IPriceObserver.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "prb-math/contracts/PRBMathSD59x18.sol";
+
+import "hardhat/console.sol";
+
+contract SMAOracle is Ownable, IOracleWrapper {
+    using PRBMathSD59x18 for int256;
+
+    /// Price oracle supplying the spot price of the quote asset
+    address public override oracle;
+
+    /// Price observer providing the SMA oracle with historical pricing data
+    address public observer;
+
+    /// Current SMA price
+    int256 public price;
+
+    /// Number of periods to use in calculating the SMA (`k` in the SMA equation)
+    uint256 public periods;
+
+    int256 public scaler;
+    uint256 public constant MAX_DECIMALS = 18;
+
+    constructor(
+        address _spotOracle,
+        uint256 _spotDecimals,
+        address _observer,
+        uint256 _periods
+    ) {
+        require(_spotOracle != address(0) && _observer != address(0), "SMA: Null address forbidden");
+        require(_periods > 0 && _periods <= IPriceObserver(_observer).capacity(), "SMA: Out of bounds");
+        require(_spotDecimals <= MAX_DECIMALS, "SMA: Decimal precision too high");
+        periods = _periods;
+        setOracle(_spotOracle);
+        setObserver(_observer);
+
+        /* `scaler` is always <= 10^18 and >= 1 so this cast is safe */
+        scaler = int256(10**(MAX_DECIMALS - _spotDecimals));
+
+        price = SMA(IPriceObserver(_observer).getAll(), _periods);
+    }
+
+    function setOracle(address _spotOracle) public override onlyOwner {
+        oracle = _spotOracle;
+    }
+
+    function setObserver(address _observer) public onlyOwner {
+        observer = _observer;
+    }
+
+    function fromWad(int256 wad) external view override returns (int256) {
+        return wad / scaler;
+    }
+
+    function getPrice() external view override returns (int256) {
+        return price;
+    }
+
+    /**
+     * @notice Add a new spot price observation to the SMA oracle
+     * @dev O(n) complexity (with n being `capacity`) due to rotation of
+     *      underlying observations array and subsequent recalculation of SMA
+     *      price
+     *
+     */
+    function update() public returns (int256) {
+        /* query the underlying spot price oracle */
+        IOracleWrapper spotOracle = IOracleWrapper(oracle);
+        int256 latestPrice = spotOracle.getPrice();
+
+        /* expire the oldest observation and load the fresh one in */
+        IPriceObserver priceObserver = IPriceObserver(observer);
+        priceObserver.add(latestPrice);
+
+        /* update current reported SMA price */
+        price = SMA(priceObserver.getAll(), periods);
+
+        return price;
+    }
```

Comment:
I feel like this function should have some sort of "limitation" on who can call it and when it will successfully update. At the moment, it seems like somebody could call `update` `n` arbitrary times very quickly and immediately change the average to whatever the current price is.